src/MPAGSCipher/PlayfairCipher.cpp

```diff
@@ -0,0 +1,173 @@
+
+    std::string cipherText = inputText;
+    int cipherSize = cipherText.size();
+    // Change J → I
+    std::transform(std::begin(cipherText), std::end(cipherText), std::begin(cipherText), [] (char x) {return (x == 'J') ? 'I' : x;});
+
+    // If repeated chars in a digraph add an X or Q if XX
+    for(int i=1; i<cipherSize;) {
+        if (cipherText[i-1] == 'X' && cipherText[i] == 'X') {
+            cipherText.insert(i, 1, 'Q');
+            i+=2;
+        } else if (cipherText[i-1] == cipherText[i]) {
+            cipherText.insert(i, 1, 'X');
+            i+=2;
+        } else {i+=2;}
+    }
+
+    // if the size of input is odd, add a trailing Z
+    if (cipherText.size() % 2 != 0) {
+        cipherText+='Z';
+    }
+
+    switch (cipherMode) {
+        case CipherMode::Encrypt:            
+            // Loop over the input in Digraphs
+            
+            // - Find the coords in the grid for each digraph
+
+            // - Apply the rules to these coords to get 'new' coords
+
+            // - Find the letter associated with the new coords
+            cipherSize = cipherText.size();
+            for(int i=1; i<cipherSize; i+=2) {
+                auto charCoord1 = (*keyMatrix.find(cipherText[i-1])).second;
+                auto charCoord2 = (*keyMatrix.find(cipherText[i])).second;
+                if (charCoord1.first == charCoord2.first) { //Checking if Rows are Equal
+                    charCoord1 = charCoord2;
+                    if (charCoord2.second == 4) {
+                        charCoord2 = std::pair<int, int>{charCoord2.first, charCoord2.second % 4};
+                    } else {
+                        charCoord2 = std::pair<int, int>{charCoord2.first, charCoord2.second + 1};
+                    }
+
+                    cipherText[i-1] = (*coordMatrix.find(charCoord1)).second;
+                    cipherText[i] = (*coordMatrix.find(charCoord2)).second;
+                } else if (charCoord1.second == charCoord2.second) {
+                    charCoord1 = charCoord2;
+                    if (charCoord2.first == 4) {
+                        charCoord2 = std::pair<int, int>{charCoord2.first % 4, charCoord2.second};
+                    } else {
+                        charCoord2 = std::pair<int, int>{charCoord2.first + 1, charCoord2.second};
+                    } 
+                    
+                    cipherText[i-1] = (*coordMatrix.find(charCoord1)).second;
+                    cipherText[i] = (*coordMatrix.find(charCoord2)).second;
+                } else {
+                    auto tempCoord = charCoord1.second;
+                    charCoord1.second = charCoord2.second;
+                    charCoord2.second = tempCoord;
+                    
+                    cipherText[i-1] = (*coordMatrix.find(charCoord1)).second;
+                    cipherText[i] = (*coordMatrix.find(charCoord2)).second;
+                }
+            }
+            break;
+
+        case CipherMode::Decrypt:
```

Comment:
You can also get away without having separate `case`'s for encrypt and decrypt as the difference is only in the sign of the shift. If you set the shift as a variable before this code then you would avoid a good chunbk of copy+paste!